### PR TITLE
Lower tuples to separate atomic nodes where possible

### PR DESCRIPTION
**Pull request by abadams**
If tuple components have no hazards between them, they can be computed
one after another in separate atomic blocks instead of in a single
combined atomic block. This lets us use atomic adds or CAS loops more
often and mutexes less often.

As a side-effect, this also means that update definitions that don't
modify one of the tuple components now entirely elide that store as a
no-op (see the change to the rfactor test).

This is part of the atomic_vectorization branch that can be separately merged.

### CODE REVIEW

**Comment by steven-johnson** [comment]
```
/home/halidenightly/build_bot/worker/linux-64-gcc53-testbranch-trunk/halide/Makefile:1930: recipe for target 'error_atomics_vectorized_mutex' failed
make: *** [error_atomics_vectorized_mutex] Error 1
make: *** Waiting for unfinished jobs....
```

**Comment by abadams** [comment]
Curious. I didn't think to test that error case locally because it's deleted in the branch I took this change from. I'll investigate...

**Comment by abadams** [comment]
A case we formerly could not compile (it required vectorized mutexes) we can now compile correctly (it does not require mutexes at all). I'll make the test harder.

**Comment by steven-johnson** [comment]
Windows failure looks unrelated; LGTM to land pending approval from @BachiLi

File: src/SplitTuples.cpp [hunk 1]
```diff
@@ -165,31 +155,173 @@ class SplitTuples : public IRMutator {
...
         }
+    }
 
-        return result;
+    Stmt visit(const Provide *op) override {
+        return visit_provide(op, nullptr);
+    }
+
+    Stmt visit(const Atomic *op) override {
+        if (const Provide *p = op->body.as<Provide>()) {
```
Comment (BachiLi) [inline src/SplitTuples.cpp:320]: Is it possible to have other nodes between the Atomic node and the Provide node? Also can we have multiple Provides inside an Atomic node?

File: test/correctness/atomic_tuples.cpp [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+        f(x, y) = {x, y, 0};
+        f(x, y) = {f(x, y)[1] + x,
+                   f(x, y)[0] + y,
+                   f(x, y)[2] + 1};
+
+        // The summation is coupled across the first two tuple
+        // components and there are two stores, so we need a mutex
+        // there. The last store could in principle be a separate atomic
+        // add, but we instead just pack it into the critical section.
```
Comment (BachiLi) [inline test/correctness/atomic_tuples.cpp:104]: is this a missing optimization opportunity that should be documented as a github issue?

File: src/SplitTuples.cpp [hunk 3]
```diff
@@ -165,31 +155,173 @@ class SplitTuples : public IRMutator {
...
+            if (atomic && !separate_atomic_nodes_per_store) {
+                s = Atomic::make(atomic->producer_name, atomic->mutex_name, s);
+            }
+            return s;
         }
+    }
 
-        return result;
+    Stmt visit(const Provide *op) override {
+        return visit_provide(op, nullptr);
```
Comment (BachiLi) [inline src/SplitTuples.cpp:316]: Maybe just `visit_provide(op)`? I don't have strong preference.

File: test/correctness/atomic_tuples.cpp (hunk 2 above)
Comment (abadams) [reply test/correctness/atomic_tuples.cpp:104]: I'm not sure that it's actually an optimization

Comment (BachiLi) [reply test/correctness/atomic_tuples.cpp:104]: ah yes i agree

File: src/SplitTuples.cpp (hunk 3 above)
Comment (abadams) [reply src/SplitTuples.cpp:316]: I removed the default arg from visit_provide to make it unambiguous

File: src/SplitTuples.cpp (hunk 1 above)
Comment (abadams) [reply src/SplitTuples.cpp:324]: Not at this point in lowering. I added a comment to that effect.

**Comment by steven-johnson** [comment]
Windows failures are unrelated, OK to land when other bots go green

**Pull request merged**